torch/testing/_internal/optests/aot_autograd.py

```diff
@@ -0,0 +1,115 @@
+import torch
+import torch.utils._pytree as pytree
+from torch.testing._internal.common_methods_invocations import wrapper_set_seed
+from functorch.compile import compiled_function, min_cut_rematerialization_partition, nop
+from .make_fx import randomize
+import re
+
+
+class assert_raises_regex:
+    def __init__(self, exception_cls, regex):
+        self.exception_cls = exception_cls
+        self.regex = regex
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, traceback):
+        if exc_type == self.exception_cls:
+            msg = str(exc_val)
+            if not re.search(self.regex, msg):
+                raise AssertionError(
+                    f"Expected exception to match regex. regex: {self.regex}, exception: {msg}")
+            return True  # Squashes the exception
+        if exc_type is not None:
+            raise AssertionError(
+                f"Expected {self.exception_cls} to be raised, instead got exception {exc_type}")
+        raise AssertionError("Expected exception to be raised but none was")
+
+
+def aot_autograd_check(
+        func,
+        args,
+        kwargs,
+        dynamic,
+        assert_raises_regex_fn=assert_raises_regex,
+        assert_equals_fn=torch.testing._comparison.assert_close,
+        try_check_data_specialization=False):
+    flat_args, args_spec = pytree.tree_flatten((args, kwargs))
+    args_is_tensor = [isinstance(arg, torch.Tensor) for arg in flat_args]
+    args = [arg for arg in flat_args if isinstance(arg, torch.Tensor)]
+
+    # We construct a new function that only accepts Tensors as inputs
+    def func_no_tensors(args):
+        reconstructed_flat_args = []
+        args = iter(args)
+        for idx, v in enumerate(flat_args):
+            if isinstance(v, torch.Tensor):
+                reconstructed_flat_args.append(next(args))
+            else:
+                reconstructed_flat_args.append(v)
+
+        c_args, c_kwargs = pytree.tree_unflatten(reconstructed_flat_args, args_spec)
+        return func(*c_args, **c_kwargs)
+
+    compiled_f = compiled_function(
+        func_no_tensors, nop, nop, dynamic=dynamic, partition_fn=min_cut_rematerialization_partition)
+    _test_aot_autograd_forwards_backwards_helper(
+        func_no_tensors, compiled_f, args, assert_raises_regex_fn, assert_equals_fn,
+        try_check_data_specialization)
+
+
+def _test_aot_autograd_forwards_backwards_helper(
+        f, compiled_f, args, assert_raises_regex_fn, assert_equals_fn,
+        try_check_data_specialization):
+    # Verify grads are equal between compiled and non-compiled versions of f.
+
+    def call_forwards_backwards(f, args):
+        flat_args, _ = pytree.tree_flatten(args)
+        diff_args = [arg for arg in flat_args if isinstance(arg, torch.Tensor) and
+                     arg.requires_grad and arg.is_leaf]
```

Comment:
Probably not a big deal, but is there any reason to have the `arg.is_leaf` check in here? Now that we're using `autograd.grad()`, we can just get gradients w.r.t. all inputs that require grad (leaf or not) and make sure they're right.

The example in my head is that we have a bunch of AOTAutograd tests that test input mutations. You can't mutate leaf tensors, so those tests create a leaf, do some compute on it, and pass that (non-leaf) computed tensor as an input to aot autograd. Those tests don't use this util though, so this probably doesn't really matter.

Reply:
Yes, you're right that we don't need the arg.is_leaf check. I wrote this code before I modified the logic to use autograd.grad instead of .backward() and didn't update it; this is another thing that is nice about switching to torch.autograd.grad for testing